### PR TITLE
[WIP] AB suite enhancement

### LINKED ISSUE #216: simplify benchmark for torchserve
There are a ton of utilities out there that use light benchmarking tools. Ex: https://github.com/awslabs/deeplearning-benchmark

### PR DESCRIPTION
**Pull request by shivamshriwas**
## Description

AB benchmark suite enhancement.
- adds python script to run AB suite
- adds support for command-line args and support to run benchmark via configuration file 
- adds support for local & docker env execution as well already running TS
- adds support for preset plans (predefined configuration templates) to benchmark 
- adds support for CSV output report and prediction latency graph.

Fixes #216 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the tests [UT/IT] that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced. 
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- UT/IT execution results

- Logs
   -  Execution of AB suite on docker gpu:
1) console logs : [cmdline_gpu_docker.txt](https://github.com/pytorch/serve/files/4842614/cmdline_gpu_docker.txt)
2) Generated CSV output: [ab_report.csv.txt](https://github.com/pytorch/serve/files/4842617/ab_report.csv.txt)
3) Latency graph: predict_latency
4) AB output: [result.txt](https://github.com/pytorch/serve/files/4842618/result.txt)

    - Execution of AB suite on docker cpu: 
[docker_cpu.zip](https://github.com/pytorch/serve/files/4842640/docker_cpu.zip)

    - Execution of AB suite on Local env gpu:
[gpu_local.zip](https://github.com/pytorch/serve/files/4842642/gpu_local.zip)

## Checklist:

- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] New and existing unit tests pass locally with these changes?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: benchmarks/ab_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        """grep "Requests per second:" {0} | awk '{1}'""".format(result_file, "{print $4}"))
+    artifacts['ts_p50'] = execute_return_stdout("""grep "50\%" {0} | awk '{1}'""".format(result_file, "{print $NF}"))
+    artifacts['ts_p90'] = execute_return_stdout("""grep "90\%" {0} | awk '{1}'""".format(result_file, "{print $NF}"))
+    artifacts['ts_p99'] = execute_return_stdout("""grep "90\%" {0} | awk '{1}'""".format(result_file, "{print $NF}"))
+    artifacts['ts_mean'] = execute_return_stdout(
+        """grep -E "Time per request:.*mean\)" {0} | awk '{1}'""".format(result_file, "{print $4}"))
+    artifacts['ts_error_rate'] = execute_return_stdout(
+        """echo "scale=2; 100*{0}/{1}" | bc | awk '{2}'""".format(int(artifacts['ts_error']),
+                                                                  execution_params['requests'],
+                                                                  "{printf \"%f\", $0}"))
```
Comment (dhanainme) [inline benchmarks/ab_benchmark.py:214]: Can we parse this with RegEX instead of calling grep in a subprocress.

File: benchmarks/ab_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+@click.option('--batch_delay', '-bd', default=200, help='batch delay')
+@click.option('--input', '-i', 'input_file', default='../examples/image_classifier/kitten.jpg',
+              type=click.Path(exists=True),
+              help='model input')
+@click.option('--workers', '-w', default=4, help='model workers')
+@click.option('--content_type', '-ic', default='application/jpg', help='content type')
+@click.option('--image', '-di', default='', help='custom docker image')
+@click.option('--docker_runtime', '-dr', default='', help='docker runtime')
+@click.option('--ts', '-ts', type=click.BOOL, default=False, help='use already running docker instace')
+@click_config_file.configuration_option(provider=json_provider, help="read configuration from a JSON file")
```
Comment (dhanainme) [inline benchmarks/ab_benchmark.py:56]: Please move this to a method invoked by __main__

File: benchmarks/ab_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    click.secho("*Executing inference performance test", fg='green')
+    ab_cmd = f"ab -c {execution_params['concurrency']}  -n {execution_params['requests']} -k -p /tmp/benchmark/input -T " \
+             f"{execution_params['content_type']} http://127.0.0.1:8080/predictions/benchmark > {result_file}"
+    execute(ab_cmd, wait=True)
+
+    click.secho("*Unregistering model ...", fg='green')
+    execute("curl -X DELETE \"http://localhost:8081/models/benchmark\"", wait=True)
+    click.secho("*Terminating Torchserve instance...", fg='green')
+    execute("torchserve --stop", wait=True)
+    click.secho("Apache Bench Execution completed.", fg='green')
```
Comment (dhanainme) [inline benchmarks/ab_benchmark.py:108]: Use python requests module instead of curl

File: benchmarks/ab_benchmark.py (hunk 1 above)
Comment (shivamshriwas) [reply benchmarks/ab_benchmark.py:220]: Changed  to use of regex.

File: benchmarks/ab_benchmark.py (hunk 3 above)
Comment (shivamshriwas) [reply benchmarks/ab_benchmark.py]: Done

**Pull request closed without merging**